src/callbacks/check_gradients.cpp

```diff
@@ -64,40 +65,44 @@ DataType compute_objective_function(model& m) {
 
 } // namespace
 
-check_gradients
-  ::check_gradients(DataType step_size,
-                                   bool verbose,
-                                   bool error_on_failure)
-  : m_step_size(step_size),
+check_gradients::check_gradients(std::set<execution_mode> modes,
+                                 DataType step_size,
+                                 bool verbose,
+                                 bool error_on_failure)
+  : m_modes(std::move(modes)),
+    m_step_size(step_size),
     m_verbose(verbose),
     m_error_on_failure(error_on_failure) {}
 
-void check_gradients::on_test_end(model *m) {
+void check_gradients::do_check_gradients(model& m) const {
 
   // Get objects from model
-  lbann_comm *comm = m->get_comm();
-  auto mode = m->get_execution_mode();
-  const auto& layers = m->get_layers();
+  auto& comm = *m.get_comm();
+  const auto mode = m.get_execution_mode();
+  const auto& layers = m.get_layers();
+
+  // Return immediately if execution mode is invalid
+  if (!m_modes.empty() && m_modes.count(mode) == 0) { return; }
 
   // Reset statistics and gradients
-  m->get_objective_function()->reset_statistics(mode);
-  for (auto&& met : m->get_metrics()) {
+  m.get_objective_function()->reset_statistics(mode);
+  for (auto&& met : m.get_metrics()) {
```

Comment:
Not really a comment on this PR, but why do we need to reset _metrics_ here?

Reply:
We previously did gradient checking at the beginning of the test phase, which caused problems with metric checking since repeated forwardpropping messes up the metric statistics. I tried fixing this by reseting the metrics, but I couldn't figure out a good way to reset the data reader.

In any case, we should expect the metric statistics to be trashed after gradient checking, so I don't think reseting them does any harm.